create_layout.py

```diff
@@ -91,45 +91,111 @@
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a file structure -- 'before' and 'after' -- 
+    return a tuple specifying which files have been added, 
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = set()
+  modified_files = set()
+  added_files = set()
+  removed_files = set()
+  for key in before_dict:
+    if key in after_dict:
+      if before_dict[key] == after_dict[key]:
+        # Matching the hashes to check if file was unchanged
+        unchanged_files.add(key)
+      else:
+        modified_files.add(key)
+    else:
+      removed_files.add(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.add(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+      sorted(removed_files))
+
+def create_material_rules(previous_link, current_link):
   """Create generic material rules (3 variants)
 
+  previous_link: the link of previous step
+  current link: the link of current step
+
   * MATCH available materials with products from previous step (links must be an
   ordered list) and
   * ALLOW available materials if it is the first step in the
   list
+  * DELETE removed materials
+
   Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  NOTE: Read header docstring for ideas for more complexity.  
+  """
 
   expected_materials = []
-
-  if index == 0:
-    for material_name in links[index].materials.keys():
-      expected_materials.append(["ALLOW", material_name])
+  unchanged_files, modified_files, _, deleted_files = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = set(unchanged_files + modified_files + deleted_files)
+
+  if previous_link is None:
+    # If current step is the initial step, we don't need to add MATCH rules.
+    # Instead, ALLOW all the existing files and DELETE deleted files.
+    for file in curr_step_materials:
+      expected_materials.append(["ALLOW", file])
+    for file in deleted_files:
+      expected_materials.append(["DELETE", file])
     expected_materials.append(["DISALLOW", "*"])
 
   else:
-    expected_materials = [
-        ["MATCH", "*", "WITH", "PRODUCTS", "FROM", links[index - 1].name]]
+    # MATCH only those that already were in the previous step
+    for file in curr_step_materials.intersection(previous_link.products):
+      expected_materials.append(
+        ["MATCH", file, "WITH", "PRODUCTS", "FROM", previous_link.name])
+
+    # Allow the rest by name
+    for file in curr_step_materials.difference(previous_link.products):
+        expected_materials.append(["ALLOW", file])
+
+    for file in deleted_files:
```

Comment:
This is not dependent on the condition, it can live independently outside it instead of twice, in each branch.

Reply:
Do you mean when we `MATCH` an artifact, we should also `ALLOW` it? I put `ALLOW` in different branch because I think we don't need to `ALLOW` an artifact if we already `MATCH` it with the artifact from the previous step.

Reply:
No I meant the DELETE rule.